source/common/conn_pool/conn_pool_base.h

```diff
@@ -14,6 +14,10 @@ namespace ConnectionPool {
 
 class ConnPoolImplBase;
 
+// A placeholder struct for whatever data a given connection pool needs to
+// successfully attach and upstream connection to a downstream connection.
+struct AttachContext {};
```

Comment:
Oh I see, here. Yeah just add a virtual destructor and the dynamic_cast will work? This is what we do for the ThreadLocal code. Up to you.